### PR TITLE
[DRAFT][TTS] Magpietts Simple API and loading audiocodec from Huggingface

### PR DESCRIPTION
**Pull request by subhankar-ghosh**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Refactored prepare_context_tensor method
- Magpietts Simple API do_tts(transcript, language, apply_TN)
- loading audiocodec from Huggingface
- Enable text_tokenizer files to be registered as artifacts, so that they can be saved in the .nemo checkpoint file instead of loading from external files.
- Move eval_config.json from nemo/collections/tts/modules/magpietts_inference to examples/tts
- eval_config.json is now just an example - which contains the dataset for CICD.
- Changed input to inference -> Now instead of taking in dataset names, the input will be a json file which contains the dataset metadata for the datasets you want to run inference on. It will run inference (and eval) on all datasets in the .json file.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/tts/models/magpietts.py [hunk 1]
```diff
@@ -2831,6 +2835,166 @@ def setup_validation_data(self, dataset_cfg):
...
+        assert (
+            self.has_baked_context_embedding
+        ), "Model does not have a baked context embedding. Please use a checkpoint with a baked context embedding."
+        # Apply text normalization if requested
+        normalized_text = transcript
+        if apply_TN:
+            try:
+                from nemo_text_processing.text_normalization.normalize import Normalizer
+
+                normalizer = Normalizer(input_case='cased', lang=language)
```
Comment (blisc) [inline nemo/collections/tts/models/magpietts.py]: What happens if we don't have TN for a specific language?

File: nemo/collections/tts/models/magpietts.py [hunk 2]
```diff
@@ -2831,6 +2835,166 @@ def setup_validation_data(self, dataset_cfg):
     def setup_test_data(self, dataset_cfg):
         self._test_dl = self._setup_test_dataloader(dataset_cfg)
 
+    def setup_dummy_text_context_in_batch(
```
Comment (blisc) [inline nemo/collections/tts/models/magpietts.py]: How much work would it be to re-factor prepare_context_tensors() to not require these dummy tensors?

File: examples/tts/magpietts_inference.py [hunk 3]
```diff
@@ -377,10 +367,10 @@ def create_argument_parser() -> argparse.ArgumentParser:
     # Dataset and output arguments
     data_group = parser.add_argument_group('Dataset and Output')
     data_group.add_argument(
-        '--datasets',
+        '--datasets_json_path',
         type=str,
         default=None,
-        help=f'Comma-separated dataset names (default: {EVALUATION_DATASETS})',
+        help='Path to dataset configuration JSON file (will process all datasets in the file)',
     )
```
Comment (blisc) [inline examples/tts/magpietts_inference.py:374]: This should no longer have a default be a required arg

File: nemo/collections/tts/modules/magpietts_inference/evaluate_generated_audio.py [hunk 4]
```diff
@@ -32,23 +32,25 @@
 
 import nemo.collections.asr as nemo_asr
 from nemo.collections.asr.metrics.wer import word_error_rate_detail
+from nemo.collections.tts.modules.utmosv2 import UTMOSv2Calculator
 from nemo.utils import logging
```
Comment (blisc) [inline nemo/collections/tts/modules/magpietts_inference/evaluate_generated_audio.py:34]: Can you do a find and replace inside of this file of all print statements and switch them to logging?

File: nemo/collections/tts/modules/magpietts_inference/evaluate_generated_audio.py [hunk 5]
```diff
@@ -32,23 +32,25 @@
 
 import nemo.collections.asr as nemo_asr
 from nemo.collections.asr.metrics.wer import word_error_rate_detail
+from nemo.collections.tts.modules.utmosv2 import UTMOSv2Calculator
```
Comment (blisc) [inline nemo/collections/tts/modules/magpietts_inference/evaluate_generated_audio.py]: We also need to make UTMOS an optional import

File: nemo/collections/tts/modules/magpietts_inference/evaluate_generated_audio.py [hunk 6]
```diff
@@ -32,23 +32,25 @@
...
 from nemo.utils import logging
 
 # Path to evalset config JSON
-EVALSET_CONFIG_PATH = Path(__file__).parent / 'evalset_config.json'
+# Get project root: from nemo/collections/tts/modules/magpietts_inference/ go up to NeMo/
+_project_root = Path(__file__).resolve().parents[5]
+EVALSET_CONFIG_PATH = _project_root / 'examples' / 'tts' / 'evalset_config.json'
```
Comment (blisc) [inline nemo/collections/tts/modules/magpietts_inference/evaluate_generated_audio.py]: We should refactor this to work the same way as examples/tts/magpietts_inference.py

File: nemo/collections/tts/models/magpietts.py [hunk 7]
```diff
@@ -2943,6 +3026,166 @@ def setup_validation_data(self, dataset_cfg):
...
+        """Setup dummy audio context tensors in the batch dictionary."""
+        # Model has baked context - create minimal dummy context tensors
+        # These will be ignored in prepare_context_tensors when baked embedding is used
+        dummy_context_codes = torch.zeros(1, self.num_audio_codebooks, 2, device=self.device, dtype=torch.long)
+        dummy_context_codes[:, :, 0] = self.context_audio_bos_id
+        dummy_context_codes[:, :, 1] = self.context_audio_eos_id
+        batch['context_audio_codes'] = dummy_context_codes
+        batch['context_audio_codes_lens'] = torch.tensor([2], device=self.device, dtype=torch.long)
+
+    def do_tts(
```
Comment (subhankar-ghosh) [inline nemo/collections/tts/models/magpietts.py:3291]: @blisc @rlangman I wrote this api do_tts currently just for the open source checkpoint (no audio context support, no text context support). Do we want to make this more general to handle audio context and text context support as well? I can write a logic to ignore them when baked context embeddings are there. Thoughts?

File: examples/tts/magpietts_inference.py (hunk 3 above)
Comment (subhankar-ghosh) [reply examples/tts/magpietts_inference.py:374]: The default is None in which case the CICD json file (also serves as an example) is loaded - the path to which is hardcoded (default value) in nemo/collectioins/tts/modules/magpietts_inference/evaluate_generated_audio.py. Do you want to remove the hardcoded path and just make it a required arg?

File: nemo/collections/tts/modules/magpietts_inference/evaluate_generated_audio.py (hunk 4 above)
Comment (subhankar-ghosh) [reply nemo/collections/tts/modules/magpietts_inference/evaluate_generated_audio.py:34]: There are no print statements in this file. I checked. Only one `pprint.pprint(avg_metrics)` to print the average metrics. Every other print statement has been replaced by logging.info in the previous PR.

File: nemo/collections/tts/models/magpietts.py (hunk 1 above)
Comment (subhankar-ghosh) [reply nemo/collections/tts/models/magpietts.py]: Added normalizer cache and checks to skip normalization in case language does not exist in the Normalizer.

File: nemo/collections/tts/modules/magpietts_inference/evaluate_generated_audio.py (hunk 5 above)
Comment (subhankar-ghosh) [reply nemo/collections/tts/modules/magpietts_inference/evaluate_generated_audio.py]: Done

File: nemo/collections/tts/modules/magpietts_inference/evaluate_generated_audio.py (hunk 6 above)
Comment (subhankar-ghosh) [reply nemo/collections/tts/modules/magpietts_inference/evaluate_generated_audio.py]: This is the default/example json file - the CICD file. I can remove this, so no default value. If a valid .json path is not provided we can throw an error, making it a required argument.

File: examples/tts/magpietts_inference.py [hunk 8]
```diff
@@ -487,11 +477,10 @@ def main():
...
     args = parser.parse_args()
 
-    # Set default datasets if not provided
-    if args.datasets is None:
-        args.datasets = EVALUATION_DATASETS
+    dataset_meta_info = load_evalset_config(args.datasets_json_path)
+    datasets = list(dataset_meta_info.keys())
 
-    datasets = args.datasets.split(",")
+    logging.info(f"Loaded {len(datasets)} datasets from {args.datasets_json_path}: {', '.join(datasets)}")
```
[the remark this answers is not in the document]
Comment (subhankar-ghosh) [reply examples/tts/magpietts_inference.py]: Removed. Added logging about filename where it is being loaded

File: nemo/collections/tts/models/magpietts.py [hunk 9]
```diff
@@ -120,13 +120,20 @@ def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
...
+
         # load codec, disable loading of loss modules not needed during inference
-        codec_model_cfg = AudioCodecModel.restore_from(cfg.get('codecmodel_path'), return_config=True)
-        if "use_scl_loss" in codec_model_cfg:
-            codec_model_cfg.use_scl_loss = False
-        codec_model = AudioCodecModel.restore_from(
-            cfg.get('codecmodel_path'), strict=False, override_config_path=codec_model_cfg
-        )
+        codec_model_path = cfg.get('codecmodel_path')
+        if codec_model_path.startswith('nvidia/'):
```
[the remark this answers is not in the document]
Comment (subhankar-ghosh) [reply nemo/collections/tts/models/magpietts.py:174]: This is NeMo's standard way of checking huggingface model name.

File: nemo/collections/tts/modules/magpietts_inference/evaluate_generated_audio.py [hunk 10]
```diff
@@ -34,21 +34,34 @@
...
+try:
+    from nemo.collections.tts.modules.utmosv2 import UTMOSv2Calculator
+
+    UTMOSV2_AVAILABLE = True
+except (ImportError, ModuleNotFoundError) as e:
+    UTMOSV2_AVAILABLE = False
+    logging.warning(
+        f"UTMOSv2Calculator not available: {e}. "
+        "UTMOSv2 metrics will be disabled. Install required dependencies to enable."
+    )
```
Comment (blisc) [inline nemo/collections/tts/modules/magpietts_inference/evaluate_generated_audio.py:47]: Can you remove the UTMOS from the requirements? And update this comment on how to install it

File: nemo/collections/tts/models/magpietts.py (hunk 7 above)
Comment (blisc) [reply nemo/collections/tts/models/magpietts.py:3291]: Eventually yes, but let's move it out of this PR

File: examples/tts/magpietts_inference.py (hunk 3 above)
Comment (blisc) [reply examples/tts/magpietts_inference.py:374]: >Do you want to remove the hardcoded path and just make it a required arg?

Yes. Is there a reason we don't want this behaviour?

File: nemo/collections/tts/models/magpietts.py [hunk 11]
```diff
@@ -84,6 +84,52 @@ class InferBatchOutput:
     headwise_cross_attention_maps: Optional[List[Any]] = None
 
 
+@dataclass
+class ContextTensorsOutput:
```
Comment (blisc) [inline nemo/collections/tts/models/magpietts.py:88]: The changes in the refactor are great. I love how we split prepare_context_tensors into appropriate helper functions now.
One comment on the use of dataclasses. As we refactor and make the code more legible, we should aim to simplify the inputs and outputs of our functions to minimize the use of dictionaries and thus dataclasses moving forward.
The dictionary return was really bad and the dataclasses are a huge step above that. I opine that the most optimal solution is to have smaller functions that return a reasonable number of outputs rather than dataclasses

Comment (rlangman) [reply nemo/collections/tts/models/magpietts.py:88]: It is definitely an improvement. Though it also highlights that the `prepare_context` method is too large right now. There is almost no situation where a helper function should have 16 output parameters, with many of them being `None` depending on the input. It looks like it could be split into 3 separate methods which prepare (1) audio context (2) text context (3) attention prior. Not a requirement for this PR, though.

File: nemo/collections/tts/models/magpietts.py [hunk 12]
```diff
@@ -395,6 +448,85 @@ def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
...
+                    try:
+                        # register_artifact handles both:
+                        # - Local paths: registers for .nemo packaging, returns absolute path
+                        # - nemo: paths: resolves to extracted file location
+                        artifact_path = self.register_artifact(
+                            f'text_tokenizers.{tokenizer_name}.g2p.phoneme_dict',
+                            phoneme_dict_path,
+                            verify_src_exists=True,
+                        )
+                        if artifact_path:
```
Comment (rlangman) [inline nemo/collections/tts/models/magpietts.py:499]: When would `register_artifact()` return None?

File: nemo/collections/tts/models/magpietts.py [hunk 13]
```diff
@@ -395,6 +448,85 @@ def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
...
+                            with open_dict(cfg):
+                                cfg.text_tokenizers[tokenizer_name].g2p.phoneme_dict = artifact_path
+                    except FileNotFoundError:
+                        logging.warning(
+                            f"phoneme_dict file not found for tokenizer '{tokenizer_name}': "
+                            f"{phoneme_dict_path}. Artifact will not be packaged in .nemo file."
+                        )
```
Comment (rlangman) [inline nemo/collections/tts/models/magpietts.py:506]: I feel like phoneme dictionary not existing should be an error. Granted, for backwards compatibility any checkpoint we use would need to be updated to have the registered artifacts.

File: nemo/collections/tts/models/magpietts.py [hunk 14]
```diff
@@ -395,6 +448,85 @@ def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
...
+
+                # Register heteronyms (or resolve nemo: path if restoring from .nemo)
+                heteronyms_path = (
+                    g2p_cfg.get('heteronyms', None)
+                    if hasattr(g2p_cfg, 'get')
+                    else getattr(g2p_cfg, 'heteronyms', None)
+                )
```
Comment (rlangman) [inline nemo/collections/tts/models/magpietts.py:513]: What are the two scenarios leading to this if/else?

File: nemo/collections/tts/models/magpietts.py [hunk 15]
```diff
@@ -1570,163 +1702,341 @@ def embed_context_text(self, context_text_tokens):
...
+            context_audio_embedded, context_text_embedded
+        )
+
+        # Interpolate between audio and text embeddings based on has_text_context flag
+        has_text_context = batch['has_text_context'].unsqueeze(-1).unsqueeze(-1).float()  # (B, 1, 1)
+        context_embedded = has_text_context * context_text_embedded + (1 - has_text_context) * context_audio_embedded
+        context_lens = (
+            batch['has_text_context'].float() * context_text_lens
+            + (1 - batch['has_text_context'].float()) * context_audio_codes_lens
+        )  # (B,)
```
Comment (rlangman) [inline nemo/collections/tts/models/magpietts.py]: It is not relevant to this PR, but why are the tensors interpolated in this way (multiplying by 0 or 1) instead of using something like `torch.where(has_text_context, text_context, audio_context)`

Comment (subhankar-ghosh) [reply nemo/collections/tts/models/magpietts.py]: Changed it to `torch.where`

**Comment by subhankar-ghosh** [comment]
fixed the UTMOS requirements and the json in magpietts_inference. Changed to a better way of using has_text_context in magpietts.py.

File: examples/tts/evalset_config.json
Comment (rfejgin) [inline examples/tts/evalset_config.json:1]: This may be a good time to remove `feature_dir`, I believe we don't need that anymore (@paarthneekhara could you confirm?) and it would simplify both the JSON file and the code. But we'd have to remove it in a few places in the code.

**Review by blisc — APPROVED** [review]

**Review by blisc — APPROVED** [review]

File: examples/tts/evalset_config.json
Comment (subhankar-ghosh) [reply examples/tts/evalset_config.json]: Do you mean from the evalset_config.json metadata file? I would not do that in this PR, because I think this might need testing of different scenarios. But you raise a good point. I did a quick search of `feature_dir` and I did not see it being used in a lot of places. So, may be with my next PR on long-form, I would remove it.
